python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -415,6 +420,212 @@ def test_torch_tensor_nccl_overlap_timed(ray_start_regular, overlap_gpu_communic
+        start = time.monotonic()
+        ref = compiled_dag.execute(i)
+        result = ray.get(ref)
+        duration = time.monotonic() - start
+        elapses.append(duration)
+        assert (
+            result
+            == [(i * num_workers, coll_shape, dtype)] * num_workers
+            + [(i, send_shape, dtype)] * num_workers
+        )
+    duration = time.monotonic() - start
+    print(f"{overlap_gpu_communication=}, {duration=}")
+    for i, elapse in enumerate(elapses):
+        print(f"iteration {i=}, {elapse=}")
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.skipif(not USE_GPU, reason="Skipping GPU Test")
+@pytest.mark.parametrize("overlap_gpu_communication", [False, True])
+def test_torch_tensor_nccl_overlap_send_future_across_actors(
+    ray_start_regular, overlap_gpu_communication
+):
+    """
+    A GPU future cannot be sent across actors. This test checks that the
+    SharedMemoryChannel waits for the future to complete before writing
+    the result.
+    """
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) >= 2
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    dtype = torch.float16
+    coll_shape = (1_000,)
+    with InputNode() as inp:
+        coll_inputs = [worker.send.bind(coll_shape, dtype, inp) for worker in workers]
+        coll_values = collective.allreduce.bind(coll_inputs)
+        coll_outputs = [
+            # Collective results are sent across actors.
+            workers[0].recv.bind(coll_values[1]),
+            workers[1].recv.bind(coll_values[0]),
+        ]
+        dag = MultiOutputNode(coll_outputs)
+
+    compiled_dag = dag.experimental_compile(
+        _overlap_gpu_communication=overlap_gpu_communication
+    )
+
+    for i in range(5):
+        ref = compiled_dag.execute(i)
+        result = ray.get(ref)
+        assert result == [(i * num_workers, coll_shape, dtype)] * num_workers
+
+    compiled_dag.teardown()
+
+    with InputNode() as inp:
+        coll_inputs = [worker.send.bind(coll_shape, dtype, inp) for worker in workers]
+        coll_values = collective.allreduce.bind(coll_inputs)
+        # Collective results are read by the driver.
+        dag = MultiOutputNode(coll_values)
+
+    compiled_dag = dag.experimental_compile(
+        _overlap_gpu_communication=overlap_gpu_communication
+    )
+
+    for i in range(5):
+        ref = compiled_dag.execute(i)
+        results = ray.get(ref)
+        expected_value = i * num_workers
+        expected = torch.ones(coll_shape) * expected_value
+        for actual in results:
+            assert torch.equal(expected, actual.cpu())
+
+
+@pytest.mark.skipif(not USE_GPU, reason="Skipping GPU Test")
+@pytest.mark.parametrize("overlap_gpu_communication", [False, True])
+def test_torch_tensor_nccl_overlap_same_future_multiple_waits(
```

Comment:
This test seems a bit complicated / unrelated compared to the stated goal? Is there a simpler test that can be run? Or a unit test?